core/capabilities/capability.go

```diff
@@ -0,0 +1,145 @@
+package capabilities
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"golang.org/x/mod/semver"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+)
+
+// CapabilityType is an enum for the type of capability.
+type CapabilityType string
+
+// CapabilityType enum values.
+const (
+	CapabilityTypeTrigger CapabilityType = "trigger"
+	CapabilityTypeAction  CapabilityType = "action"
+	CapabilityTypeReport  CapabilityType = "report"
+	CapabilityTypeTarget  CapabilityType = "target"
+)
+
+// IsValid checks if the capability type is valid.
+func (c CapabilityType) IsValid() error {
+	switch c {
+	case CapabilityTypeTrigger,
+		CapabilityTypeAction,
+		CapabilityTypeReport,
+		CapabilityTypeTarget:
+		return nil
+	}
+
+	return fmt.Errorf("invalid capability type: %s", c)
+}
+
+// Validatable is an interface for validating the config and inputs of a capability.
+type Validatable interface {
+	ValidateConfig(config values.Map) error
+	ExampleOutput() values.Value
+	ValidateInput(inputs values.Map) error
+}
+
+// Executable is an interface for executing a capability.
+type Executable interface {
+	// Start will be called when the capability is loaded by the application.
+	// Start will be called before the capability is added to the registry.
+	Start(ctx context.Context, config values.Map) (values.Value, error)
+	// Capability must respect context.Done and cleanup any request specific resources
+	// when the context is cancelled. When a request has been completed the capability
+	// is also expected to close the callback channel.
+	// Request specific configuration is passed in via the inputs parameter.
+	Execute(ctx context.Context, callback chan values.Value, inputs values.Map) error
```

Comment:
I left this question in the CRON doc: if Execute() registers a trigger, how do we de-register it when a workflow is deleted?

Reply:
This is the responsibility of a workflow engine. There are 2 steps to capabilities:
- A capability is registered by a NOP.
- A capability is `start`ed when a workflow requiring it is received.
(I can already see the comment on line 47 being wrong 😅)

Similarly, when a workflow requiring a capability is removed, the capability should be `stop`ed. @cedric-cordenier we'll need to add more details here.

Reply:
OK, if start/stop is per workflow then it makes more sense.